text/0000-router-service.md

```diff
@@ -0,0 +1,252 @@
+
+### New Methods: Initiating Transitions
+
+```js
+transitionTo(routeName, ...models, queryParams)
+replaceWith(routeName, ...models, queryParms)
+```
+
+These two have the same semantics as the existing methods on `Ember.Route`:
+
+### New Method: Checking For Active Route
+
+ - `isActive(routeName, ...models, queryParams)`
+
+The arguments have the same semantics as `transitionTo`, the return value is a boolean. This should provide the same logic that determines whether to put an active class on a `link-to`. Here's an example of how we can implement `is-active` as a helper, using this method:
+
+```js
+import Helper from 'ember-helper';
+import service from 'ember-service/inject';
+import observer from 'ember-metal/observer';
+
+export default Helper.extend({
+  router: service(),
+  compute([routeName, ...models], hash) {
+    let allModels;
+    if (hash.models) {
+      allModels = models.concat(hash.models);
+    } else {
+      allModels = models;
+    }
+    return this.get('router').isActive(routeName, ...models, hash.queryParams);
+  },
+  observer('router.currentRoute', function() {
+    this.recompute();
+  });
+});
+```
+
+```hbs
+{{!- Example usage -}}
+<li class={{if (is-active "person.detail" model) 'chosen'}} >
+
+{{!- Example usage with generic routeName and list of models (avoids splat) -}}
+<a class={{if (is-active routeName models=models) 'chosen'}} >
+
+{{!- Note that the complexities of currentWhen can be avoided by composing instead. }}
+<a class={{if (or (is-active 'one') (is-active 'two')) 'active'}} href={{url-for 'two'}} >
+
+```
+
+
+### New Method: URL generation
+
+`url(routeName, ...models, queryParams)`
+
+This takes the same arguments as `transitionTo`, but instead of initiating the transition it returns the resulting URL as a string.
+
+A `url-for` helper can be implemented almost identically to the `is-active` example above.
+
+
+### New Properties
+
+`currentRoute`: an observable property. It is guaranteed to change whenever a route transition happens (even when that transition only changes parameters and doesn't change the active route). You should consider its value deeply immutable -- we will replace the whole structure whenever it changes. The value of `currentRoute` is a `RouteInfo` representing the current leaf route. `RouteInfo` is described below.
+
+`currentRouteName`:  a convenient alias for `currentRoute.name`.
+
+`currentURL`: provides the serialized string representing `currentRoute`.
+
+
+### Deprecation
+
+I propose deprecating the publicly extensible `willTransition` and `didTransition` hooks. They are redundant with an observable `currentRoute`, and the arguments they receive leak internal implemetation from router.js.
+
+## RouteInfo Type
+
+A RouteInfo object has the following properties. They are all read-only.
+
+ - name: the dot-separated, fully-qualified name of this route, like `"people.index"`.
```

Comment:
Having the name is great but do you think it would be worth giving the instance or the class as well?  I would imagine you could use ember with something that allows you to specify "query fragments" (Falcor) in your routes and potentially components if you have a way of collecting that fragments recursively.

For example you could imagine that in node you could do something like:

```js
  let indexHTML = fs.readFileSync(__dirname+'/index.html').toString();
  const dataRegex = /%DATA%/;

  const buildQuery = (handler, query = []) => {
    query = query.concat(handler.instance.collectQuery());
    if (handler.parent) {
      buildQuery(handler.parent, query)
    }
    return callUpstreamService(query)
  }
  app.colllectQueryFor('/profile/123').then(buildQuery).then(data => {
     let indexHTML = indexHTML.replace(dataRegex, JSON.stringify(data));
     res.writeHead(200, {
       'Content-Length': indexHTML.length,
       'Content-Type': 'text/html'
     });
     res.write(indexHTML);
     res.end();
  })
```
 
This would allow you to extract the data requirements without a lot of overhead. If not, this can be proved out in user land as you should be able to lookup the route instances from the info.